sbi/neural_nets/embedding_nets.py

```diff
@@ -268,17 +273,32 @@ def forward(self, x: Tensor) -> Tensor:
         """
         batch, permutation_dim, _ = x.shape
 
-        iid_embeddings = self.trial_net(x.view(batch * permutation_dim, -1)).view(
-            batch, permutation_dim, -1
-        )
+        # if no NaNs for padding varying trial lengths we can batch the computation
+        if not torch.isnan(x).any():
+            trial_embeddings = self.trial_net(x.view(batch * permutation_dim, -1)).view(
+                batch, permutation_dim, -1
+            )
+            combined_embedding = self.combining_function(trial_embeddings, dim=1)
+            trial_counts = torch.ones(batch, 1, dtype=torch.float32) * permutation_dim
 
-        if self.combining_operation == "mean":
-            e = iid_embeddings.mean(dim=1)
-        elif self.combining_operation == "sum":
-            e = iid_embeddings.sum(dim=1)
+        # otherwise we need to loop over the batch to account for varying trial lengths
```

Comment:
shouldn't it also work to embed the whole batch and only mask out the nans after the embedding?
for the trial net, the single trials are embedded independently and it nans in one trial should not interact with nans in another trial. this would avoid looping.

Reply:
in principle yes, but in practice no. It will result in all NNs to be `NaN` after the first optimization step. I haven't figured out why yet.

Reply:
I digged a bit deeper and it seems the embedding net outputs are slightly different when calculating them using a loop vs. when passing a batch, e.g., small numerical difference around `1e-8`. not sure that's the reason. 

But even if I calculate the forward pass using both approaches on every call and then only return the batched one it fails, while it works fine when returning the looped one.